### PR TITLE
Fix #312589: Frames with text are not accessible

### PR DESCRIPTION
**Pull request by HemantAntony**
Resolves: [#312589](https://musescore.org/en/node/312589)

- [x] I signed [CLA](https://musescore.org/en/cla)
- [x] I made sure the code in the PR follows [the coding rules](https://github.com/musescore/MuseScore/wiki/CodeGuidelines)
- [x] I made sure the code compiles on my machine
- [x] I made sure there are no unnecessary changes in the code
- [x] I made sure the title of the PR reflects the core meaning of the issue you are solving
- [x] I made sure the commit message(s) contain a description and answer the question "Why do those changes fix that particular issue?" or "Why are those changes really necessary as improvements?"
- [x] I made sure the commit message title starts with "fix #424242:" if there is a related issue
- [ ] I created the test (mtest, vtest, script test) to verify the changes I made

### CODE REVIEW

**Comment by HemantAntony** [comment]
@shoogle This PR solves [#312589](https://musescore.org/en/node/312589)

**Comment by shoogle** [comment]
@HemantAntony, thanks for the fix! This works well for Vertical Frames that contain Title, Subtitle, Composer and Lyricist text, but it is possible for these frames to contains other text styles as well (or it would be except the feature is broken in MU4; I'll create an issue about that separately).

## Example Score created in MU3
- [Frame Navigation.zip](https://github.com/musescore/MuseScore/files/8878493/Frame.Navigation.zip) (you need to change the extension to `.mscz` because of GitHub)

The file demonstrates more complicated use of frames with text, including:

### 1. Text Frames with different text styles
image

### 2. Horizonal Frame with multiple instances of the Frame style
image

### 3. Vertical Frame with custom styles and multiple instances of each style
image

## Desired behaviour

Your code already handles (1) correctly: press Alt+Right to select the Text Frame, then press Alt+Right again to select the text element inside.

For (2) and (3), you need to make sure that Alt+Right is able to get to every text element in the Horizonal / Vertical Frame. At the moment it skips elements that don't use the Title, Subtitle, Composer or Lyricist style, and it only accesses the last instance of that style (e.g. "Title 2" but not "Title 1").

The order in which text elements are accessed is not important at this stage, so I suggest you just access them in whatever order the elements appear in the frame's list of children. The screen reader will say "Title" or "Composer" before it announces the actual title or composer name, so it doesn't matter if the order is different sometimes because the user added the composer before they added the title, etc.

Whatever the order is for Alt+Right, make sure that Alt+Left accesses all the elements in the reverse of that order.

**Comment by shoogle** [comment]
As discussed on Discord, @HemantAntony is going to attempt to traverse text elements in order based on their position on the page, starting in the top left and working towards the bottom right. I've updated the example file in the [comment above](https://github.com/musescore/MuseScore/pull/11951#issuecomment-1152295522) to add "Composer 3" and "Lyricist 3" at the exact same positions as "Composer 2" and "Lyricist 2" respectively. This is just to make sure that they are all traversed by Alt+Right. The exact order (2 then 3, or 3 then 2) is not important.
image

File: src/engraving/libmscore/navigate.cpp [hunk 1]
```diff
@@ -760,8 +818,61 @@ EngravingItem* Score::prevElement()
...
+            for (EngravingObject* object : list) {
+                EngravingItem* engravingItem = toEngravingItem(object);
+
+                if (!engravingItem) {
+                    continue;
+                }
+
+                boxChildren.insert(std::pair<EngravingItem*, int>(engravingItem, index));
+                index++;
+            }
```
Comment (shoogle) [inline src/engraving/libmscore/navigate.cpp:867]: Most of this code is identical to `nextElement()` so it should be factored out into a separate `static` function that takes the box as an argument and returns the set of child+index pairs. You can use the following declarations to make the code neater:

```c++
using ElementPair = std::pair<EngravingItem*, int>; // element and its index in parent

static bool compareElementPairs(const ElementPair& child1, const ElementPair& child2)
{
    PointF p1 = ...;
    PointF p2 = ...;
    // etc.
}

static std::set<ElementPair, decltype(&compareFunction)> childPairsSet(const EngravingItem* element)
{
    std::set<ElementPair, decltype(&compareFunction)> children(&compareElementPairs);

   // add elements to set

   return children;
}
```

Put that somewhere near the top of `navigate.cpp` outside of any functions, then inside `nextElement()` and `previousElement()` you just have to do:

```c++
auto boxChildren = childPairsSet(e);
// now use the set
```

**Review by shoogle — CHANGES_REQUESTED** [review]
I've suggested some code improvements but the implementation works very well indeed!

File: src/engraving/libmscore/navigate.cpp [hunk 2]
```diff
@@ -695,6 +766,26 @@ EngravingItem* Score::nextElement()
         case ElementType::VBOX:
         case ElementType::HBOX:
         case ElementType::TBOX: {
+            auto boxChildren = toChildPairsSet(e);
+
+            if ((getSelectedElement()->type() == ElementType::VBOX
```
Comment (Eism) [inline src/engraving/libmscore/navigate.cpp:771]: Please move getSelectedElement() to a variable and use it

**Comment by HemantAntony** [comment]
@Eism I updated the PR

**Review by Eism — APPROVED** [review]

File: src/engraving/libmscore/navigate.cpp [hunk 3]
```diff
@@ -824,6 +930,20 @@ EngravingItem* Score::prevElement()
         case ElementType::VBOX:
         case ElementType::HBOX:
         case ElementType::TBOX: {
+            auto boxChildren = toChildPairsSet(e);
+
+            for (auto child = boxChildren.rbegin(); child != boxChildren.rend(); child++) {
+                if (getSelectedElement() != child->first) {
```
Comment (shoogle) [inline src/engraving/libmscore/navigate.cpp:936]: @HemantAntony, you should also store `getSelectedElement()` in a variable outside the loop here.

Comment (HemantAntony) [reply src/engraving/libmscore/navigate.cpp]: Oh right, I saw it in only one place and forgot that it was in a loop

**Review by shoogle — APPROVED** [review]

**Comment by DmitryArefiev** [comment]
Tested [#312589](https://musescore.org/en/node/312589) on Win10, Mac12, LinuxMint 20.2 - **FIXED** (now is able to navigate with Alt+arrows between different text inside one frame, e.g. Compose - Subtitle - Title)

Also there is annoyance bug when switching between text elements after editing #11214 which also effects editing process inside text frame

**Pull request merged**